File: depth-first-search-medium/130.被围绕的区域.cpp
/*
 * @lc app=leetcode.cn id=130 lang=cpp
 *
 * [130] 被围绕的区域
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    int row, col;
    int dirt[4][2] = {{-1, 0}, {1, 0}, {0, 1}, {0, -1}};
    void dfs(vector<vector<char>> &board, int r, int c) {
        board[r][c] = '#';
        for (int i = 0; i < 4; i++) {
            int x = r + dirt[i][0], y = c + dirt[i][1];
            if (-1 < x && x < row && -1 < y && y < col && board[x][y] == 'O') {
                board[x][y] = '#';
                dfs(board, x, y);
            }
        }
    }
    void solve(vector<vector<char>> &board) {
        row = board.size(), col = board.size() == 0 ? 0 : board[0].size();
        if (row == 0 || col == 0) {
            return;
        }
        for (int i = 0; i < row; i++) {
            if (board[i][0] == 'O')
                dfs(board, i, 0);
            if (board[i][col - 1] == 'O')
                dfs(board, i, col - 1);
        }
        for (int i = 0; i < col; i++) {
            if (board[0][i] == 'O')
                dfs(board, 0, i);
            if (board[row - 1][i] == 'O')
                dfs(board, row - 1, i);
        }
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (board[i][j] == 'O') board[i][j] = 'X';
                if (board[i][j] == '#') board[i][j] = 'O';
            }
        }
    }
};
// @lc code=end

File: depth-first-search-medium/130.被围绕的区域.py
#
# @lc app=leetcode.cn id=130 lang=python3
#
# [130] 被围绕的区域
#
from typing import List


# @lc code=start
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        row, col = len(board), 0 if len(board) == 0 else len(board[0])

        def dfs(r: int, c: int) -> None:
            board[r][c] = '#'
            for i, j in [(-1, 0), (0, 1), (1, 0), (0, -1)]:
                x, y = r + i, c + j
                if -1 < x < row and -1 < y < col and board[x][y] == 'O':
                    board[x][y] = '#'
                    dfs(x, y)

        for i in range(0, row):
            if board[i][0] == 'O':
                dfs(i, 0)
            if board[i][col - 1] == 'O':
                dfs(i, col - 1)

        for j in range(0, col):
            if board[0][j] == 'O':
                dfs(0, j)
            if board[row - 1][j] == 'O':
                dfs(row - 1, j)

        for i in range(0, row):
            for j in range(0, col):
                if board[i][j] == '#':
                    board[i][j] = 'O'
                elif board[i][j] == 'O':
                    board[i][j] = 'X'
# @lc code=end
